### PR TITLE
Remove custom [[HasInstance]] behavior from interface objects

### LINKED ISSUE #129: Should probably spec a hasInstance for DOM interface objects
Right now we have a [[HasInstance]] thing at http://heycam.github.io/webidl/#es-interface-hasinstance but that's gone in ES6.

Proposed prose with similar semantics is something like:
1. Let `F` be the `this` value.
2. Let `V` be the first argument we were passed.
3. If `F` is not an object, return `false`.
4. If `F` is not an interface object, return `false` (or delegate to `OrdinaryHasInstance`?)
5. If `V` is not an object, return `false`.
6. If `V` is an object that implements the interface for which `F` is  the interface object (defined in terms of brands, whatever), return `true`.
7. Return `OrdinaryHasInstance(F, V)`.

@domenic, thoughts?

An open question is whether `Node[Symbol.hasInstance] == Element[Symbol.hasInstance]` should test true or false.  We could go either way, given the above prose, since they would implement the same algorithm...

### PR DESCRIPTION
**Pull request by domenic**
This was only ever implemented in Firefox. It was added in 5982ce7e323cffd7a673cb9af973c1c4f1e76fbe. after being removed as a result of https://www.w3.org/Bugs/Public/show_bug.cgi?id=12295.

Closes #129.

---

This ensures that authors can maintain the invariant they are used to where `x instanceof Y` means that `Y.prototype`'s methods are available on `x` (which is not true in Firefox). It also aligns with normal author-created or built-in ECMAScript objects, where e.g. `{} instanceof otherWindow.Object` returns false.

As pointed out in #129, only @travisleithead from a non-Firefox browser expressed support for the currently-specced behavior, and they have not implemented it in the intervening 4 years; Travis, I'd be curious if your team's feelings have changed. Chrome strongly objects to the currently-specced behavior. /cc @cdumez (EDIT: I guess @samweinig is probably the right person for Web IDL) to hear about WebKit's opinion.

As such, we'd like to remove this from the spec, as it seems that it never got consensus for being added, although people may not have voiced their objections in a timely manner when it was originally proposed.

---

If there is a desire for adding a brand-checking API to the platform, we'd like that to be proposed separately. But simply as a matter of matching reality and improving interop, we think it's important to get the spec fixed on this matter sooner, and not block on such an API addition.

***
[Preview](https://s3.amazonaws.com/pr-preview/heycam/webidl/hasinstance-non-exotic.html) | [Diff](https://s3.amazonaws.com/pr-preview/heycam/webidl/9d039f6...0881b62.html)

### CODE REVIEW

**Comment by cdumez** [comment]
I support this change. I think it is best to align with EcmaScript here.

**Comment by annevk** [comment]
Do you want me to submit a WPT for https://github.com/heycam/webidl/issues/129#issuecomment-292522793 with the opposite result? Or are you taking care of it?

File: index.bs [hunk 1]
```diff
@@ -6384,9 +6383,8 @@ more of the following are redefined in accordance with the rules for exotic obje
 \[[Call]],
 \[[Set]],
 \[[DefineOwnProperty]],
-\[[GetOwnProperty]],
-\[[Delete]] and
-\[[HasInstance]].
```
Comment (tobie) [inline index.bs]: We're still [modifying [[HasInstance]]](https://heycam.github.io/webidl/#es-legacy-callback-interface-object-hasinstance) for legacy callback interfaces, so you might want to keep it listed here.

**Comment by domenic** [comment]
I can take care of the tests; I wanted to make sure this would be merged first.

**Comment by domenic** [comment]
It appears nobody implements the throwing [[HasInstance]] for NodeFilter. We should probably just remove that.

**Comment by tobie** [comment]
> It appears nobody implements the throwing [[HasInstance]] for NodeFilter. We should probably just remove that.

Why is it even specified?

**Comment by domenic** [comment]
Unclear. Maybe because it's not really a sensible thing to be testing. But per spec it will throw anyway in most cases because these objects have no `.prototype`; see forthcoming tests.

**Comment by domenic** [comment]
@bzbarsky OK to merge this?

**Comment by bzbarsky** [comment]
Yes.  Note that Firefox is probably not going to be able change its behavior for a while; certainly not until we're able to offer a different branding-detection solution to consumers of the current behavior...

**Comment by annevk** [comment]
@bzbarsky maybe we could experiment with a chrome-only-branding API? Presumably we don't need it for web sites...

**Comment by bzbarsky** [comment]
We could try.  It'd be a bunch of work and compat risk that I'm not willing to spend time on any time in the next 6 months, personally, so just trying to set realistic expectations here.

**Comment by domenic** [comment]
Bugs commented on:

- https://bugzilla.mozilla.org/show_bug.cgi?id=1360715#c8
- https://bugs.chromium.org/p/chromium/issues/detail?id=717016#c5

Bugs for legacy callback interface objects (not directly related to this PR, but caught while writing tests in https://github.com/w3c/web-platform-tests/pull/5762):

- https://bugs.chromium.org/p/chromium/issues/detail?id=569301#c46 
- https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/11894121/

If someone could check http://w3c-test.org/WebIDL/ecmascript-binding/legacy-callback-interface-object.html in Safari Tech Preview and file a bug if necessary, I'd much appreciate it, as I'm away from my testing Mac for the next week.

**Comment by annevk** [comment]
Safari TP passes all 8 tests.

**Pull request merged**